### ISSUE TITLE
Clarification of lambda coverage by in google  style

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4006]
Based on the discussion at https://github.com/checkstyle/checkstyle/issues/3755#issuecomment-285851574 and according to [Google Java Style Guide (3 November 2016 version) ](https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R422) to cover the rule from the guide NoLineWrap should validate lambdas. It means that LAMBDA should be included in a set of acceptable tokens and the check needs to treat LAMBDAS accordingly to Google Guide by default (as we cannot predict how users will wrap lambdas lets follow Google Guide by default).

Examples of expected behavior:

```java
(String label, Long value, Object obj) -> { // OK
    ...
}

Predicate<String> predicate = str -> // OK
    isReversed(str);

(String label) -> // violation (should not have new line next to the arrow in a lambda)
{
    ...
}
```

**Comment by romani** [comment]
http://checkstyle.sourceforge.net/config_whitespace.html#NoLineWrap 
is for simple line wrapping for whole structure like package declaration and import declaration.

Here we have request to certain rules of wrapping in certain place. Even current [OperatorWrap](http://checkstyle.sourceforge.net/config_whitespace.html#OperatorWrap) is not enough.

This guide point should be covered by combination of Checks (I am not 100% sure that will satisfy) :
- NeedBrace (allowSingleLine), beware of #3837 .
- RightCurly 

@MEZk , please confirm that we are on the same position for this issue.

**Comment by MEZk** [comment]
@romani 
In the [current implementation](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoLineWrapCheck.java#L98-L98) of the check it is possible to use it for lambdas. For example:

```java
public class InputNoLineWrap1 {
    Predicate<String> isPolindrone = x ->
    {
        return foo1(1);
    };

    boolean foo1(int a) {
        return true;
    }
}
```

[getLastChild](https://github.com/MEZk/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoLineWrapCheck.java#L98-L98) will return "{" (SLIST token) and we will be able to validate whether it is on the same line as lambda or not. 

> NeedBrace (allowSingleLine), beware of #3837 .
RightCurly

I do not understand how NeedBrace and RightCurly can help us to follow the rule.

**Comment by romani** [comment]
@rdiachenko , please investigate what we can do and how it affects coverage.

**Comment by rdiachenko** [comment]
I quickly tried to add the use case from the issue description, works as expected (https://github.com/rdiachenko/checkstyle/commit/5e2cd59891b8ca43d208c497e4f4fdd7ab21eac6) when I run `WhereToBreakTest`. We need to extend inputs with cases for lamdas, as for now I don't see they are covered by the inputs for `4.5.1 Where to break` section.

**Comment by romani** [comment]
@rdiachenko , please send PR with Input extensions as you already have code

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/pull/15507 is merged, do we need to improve in Check something ?

**Comment by romani** [comment]
I do not understand this issue.

usage in style guide:

[3.2 Package statement](https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s3.2-package-statement)  
[NoLineWrap](https://checkstyle.org/checks/whitespace/nolinewrap.html#NoLineWrap)

[3.3.2 No line-wrapping](https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s3.3.2-import-line-wrapping)  NoLineWrap](https://checkstyle.org/checks/whitespace/nolinewrap.html#NoLineWrap)


all this sections are not about lambdas.

----------


issue is created to cover "lambda" appearance in style guide.
So lets just focus on point: do we cover all requirements that style guide has about lambdas?

We have few references about lambda in style guide:

> 4.5.1 Where to break
A line is never broken adjacent to the arrow in a lambda, except that a break may come immediately after the arrow if the body of the lambda consists of a single unbraced expression. Examples:
MyLambda<String, Long, Object> lambda =
    (String label, Long value, Object obj) -> {
        ...
    };
Predicate<String> predicate = str ->
    longExpressionInvolving(str); 

> 4.6.2 Horizontal whitespace
the arrow in a lambda expression: (String str) -> str.length()

**Comment by romani** [comment]
all validation is LeftCurly:
```
rivanov@p5510:/var/tmp$ java -jar checkstyle-10.17.0-all.jar -c /google_checks.xml /home/rivanov/java/github/romani/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule451wheretobreak/InputLambdaBodyWrap.java
Starting audit...
[WARN] /home/rivanov/java/github/romani/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule451wheretobreak/InputLambdaBodyWrap.java:69:11: 
    '{' at column 11 should be on the previous line. [LeftCurly]
[WARN] /home/rivanov/java/github/romani/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule451wheretobreak/InputLambdaBodyWrap.java:76:11:
     '{' at column 11 should be on the previous line. [LeftCurly]
[WARN] /home/rivanov/java/github/romani/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule451wheretobreak/InputLambdaBodyWrap.java:78:15:
   '{' at column 15 should be on the previous line. [LeftCurly]
Audit done.

```

**Comment by romani** [comment]
I think we need to close this issue as WONTFIX, lambda is fully covered, no worries. NoLineWrap completely unrelated in this context.

**Comment by rdiachenko** [comment]
> So lets just focus on point: do we cover all requirements that style guide has about lambdas?

We have enough coverage for lambdas in 4.5.1 and 4.6.2. Let's close the issue after https://github.com/checkstyle/checkstyle/pull/15510 is merged.
